[Lockdown Mode] Make sure Accept Header matches LDM capabilities

brentfulgham opened the pull request:
#### f8332133a12b758485dd348da4c6ed991bde98f5
<pre>
[Lockdown Mode] Make sure Accept Header matches LDM capabilities
<a href="https://bugs.webkit.org/show_bug.cgi?id=293385">https://bugs.webkit.org/show_bug.cgi?id=293385</a>
&lt;<a href="https://rdar.apple.com/problem/151333451">rdar://problem/151333451</a>&gt;

Reviewed by Pascoe.

We recently discovered that in Lockdown Mode WebKit generates an AcceptHeader containing image
types that are disabled. There is no security issue from this, but it&apos;s annoying to LDM users
who may end up downloading unusable images, etc.

This patch causes the Accept Header logic to check for Lockdown Mode state, and only emit
relevant types when communicating with a secure server.

This patch also updates WebKitTestRunner with the ability to activate Lockdown Mode for tests in
the &apos;lockdown-mode&apos; directory.

It also disambiguates checks for the enablement of the Lockdown Mode feature from the specific
use of the Lockdown Mode framework, which only exists on some systems at present. This allows
Open Source builds to activate the feature in WebKit and test it.

Tests:
    LayoutTests/http/tests/lockdown-mode/heic-shown-in-lockdown-mode.http.html
    LayoutTests/http/tests/lockdown-mode/no-heic-in-lockdown-mode.https.html

* Source/WTF/wtf/PlatformHave.h: Disambiguate the Lockdown Mode feature from the Lockdown Mode
framework.
* Source/WTF/wtf/URL.cpp:
(WTF::URL::protocolIsSecure const): Added.
* Source/WTF/wtf/URL.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::updateRequestAndAddExtraFields): Pass new argument.
* Source/WebCore/loader/cache/CachedResourceRequest.cpp:
(WebCore::acceptHeaderValueForImageResource): Only send Lockdown-supported image types
to secure servers when in lockdown mode.
(WebCore::CachedResourceRequest::acceptHeaderValueFromType):
(WebCore::CachedResourceRequest::setAcceptHeaderIfNone):
* Source/WebCore/loader/cache/CachedResourceRequest.h:
* Tools/TestRunnerShared/TestFeatures.cpp:
(WTR::shouldEnableLockdownMode): Add Lockdown Mode test feature.
(WTR::hardcodedFeaturesBasedOnPathForTest): Add check for &apos;lockdown-mode&apos; directory.
* Tools/TestWebKitAPI/Tests/WTF/URL.cpp:
(TestWebKitAPI::TEST_F(WTF_URL, ProtocolIsSecure)):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::lockdownModeEnabled const):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::configureWebpagePreferences): Set lockdown mode when appropriate.

Canonical link: <a href="https://commits.webkit.org/296345@main">https://commits.webkit.org/296345@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/40ff5fda0ac6c1ada29143321f4114630ffa3a56

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108236 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27898 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18319 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113446 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110199 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28596 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36449 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82182 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111184 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22662 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97501 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62614 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22075 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15637 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58179 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/100817 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92028 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15693 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116570 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/106796 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35299 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25989 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91223 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35672 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93776 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91008 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23193 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35897 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13661 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31069 "Build was cancelled. Recent messages:Printed configuration") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35199 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40746 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/131082 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34929 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35585 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38284 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36597 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/WebCore/loader/cache/CachedResourceRequest.cpp:
```diff
@@ -160,21 +164,33 @@ static String acceptHeaderValueForImageResource()
+#if HAVE(LOCKDOWN_MODE_FRAMEWORK)
+    bool limitToLockdownModeSet = usingSecureProtocol && PAL::isLockdownModeEnabledForCurrentProcess();
+#else
+    static bool limitToLockdownModeSet = false;
```
I think this won't work for tests on EWS.

cdumez commented on LayoutTests/http/tests/lockdown-mode/heic-shown-in-lockdown-mode.http-expected.txt:
```diff
@@ -0,0 +1,11 @@
+
+PASS successfullyParsed is true
+
+TEST COMPLETE
```
This test is likely not written correctly as `TEST COMPLETE` is printed *before* the test lines.

cdumez replied:
Usually, such tests end up being flaky on the bots.

cdumez commented on LayoutTests/http/tests/lockdown-mode/no-heic-in-lockdown-mode.https-expected.txt:
```diff
@@ -0,0 +1,12 @@
+
+PASS successfullyParsed is true
+
+TEST COMPLETE
```
Same issue

brentfulgham committed:
[Lockdown Mode] Make sure Accept Header matches LDM capabilities
https://bugs.webkit.org/show_bug.cgi?id=293385
<rdar://problem/151333451>

Reviewed by Pascoe.

We recently discovered that in Lockdown Mode WebKit generates an AcceptHeader containing image
types that are disabled. There is no security issue from this, but it's annoying to LDM users
who may end up downloading unusable images, etc.

This patch causes the Accept Header logic to check for Lockdown Mode state, and only emit
relevant types when communicating with a secure server.

This patch also updates WebKitTestRunner with the ability to activate Lockdown Mode for tests in
the 'lockdown-mode' directory.

It also disambiguates checks for the enablement of the Lockdown Mode feature from the specific
use of the Lockdown Mode framework, which only exists on some systems at present. This allows
Open Source builds to activate the feature in WebKit and test it.

Tests:
    LayoutTests/http/tests/lockdown-mode/heic-shown-in-lockdown-mode.http.html
    LayoutTests/http/tests/lockdown-mode/no-heic-in-lockdown-mode.https.html

* Source/WTF/wtf/PlatformHave.h: Disambiguate the Lockdown Mode feature from the Lockdown Mode
framework.
* Source/WTF/wtf/URL.cpp:
(WTF::URL::protocolIsSecure const): Added.
* Source/WTF/wtf/URL.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::updateRequestAndAddExtraFields): Pass new argument.
* Source/WebCore/loader/cache/CachedResourceRequest.cpp:
(WebCore::acceptHeaderValueForImageResource): Only send Lockdown-supported image types
to secure servers when in lockdown mode.
(WebCore::CachedResourceRequest::acceptHeaderValueFromType):
(WebCore::CachedResourceRequest::setAcceptHeaderIfNone):
* Source/WebCore/loader/cache/CachedResourceRequest.h:
* Tools/TestRunnerShared/TestFeatures.cpp:
(WTR::shouldEnableLockdownMode): Add Lockdown Mode test feature.
(WTR::hardcodedFeaturesBasedOnPathForTest): Add check for 'lockdown-mode' directory.
* Tools/TestWebKitAPI/Tests/WTF/URL.cpp:
(TestWebKitAPI::TEST_F(WTF_URL, ProtocolIsSecure)):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::lockdownModeEnabled const):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::configureWebpagePreferences): Set lockdown mode when appropriate.

Canonical link: https://commits.webkit.org/296345@main
